qa/rpc-tests/quality_voiding.py

```diff
@@ -0,0 +1,362 @@
+        self.sync_all()
+        assert_equal(True, cert_3_epoch_0 in self.nodes[0].getrawmempool())
+
+        mined = self.nodes[0].generate(1)[0]
+        self.sync_all()
+
+        # Check balance
+        bwt_address = self.nodes[1].getrawcertificate(cert_3_epoch_0, 1)['vout'][1]['scriptPubKey']['addresses'][0]
+        res = self.nodes[1].listtxesbyaddress(bwt_address)
+        assert_equal(res[0]['scid'], scid)
+        mark_logs("Height : {}".format(res[0]['vout'][1]['maturityHeight']), self.nodes, DEBUG_MODE)
+        ud = self.nodes[1].getunconfirmedtxdata(bwt_address)
+        assert_equal(bwt_amount_3, ud['bwtImmatureOutput'])
+        mark_logs("Immature output : {}".format(ud['bwtImmatureOutput']), self.nodes, DEBUG_MODE)
+
+        # Checking all certificate in mempool
+        mark_logs("Check certificates is not in mempool anymore", self.nodes, DEBUG_MODE)
+        assert_equal(False, cert_1_epoch_0 in self.nodes[0].getrawmempool())
+        assert_equal(False, cert_2_epoch_0 in self.nodes[0].getrawmempool())
+        assert_equal(False, cert_3_epoch_0 in self.nodes[0].getrawmempool())
+
+        height = self.nodes[0].getblockcount()
+        mark_logs("Check certificates in chain", self.nodes, DEBUG_MODE)
+        assert_true(cert_1_epoch_0 in self.nodes[0].getblock(self.nodes[0].getblockhash(height - 1), True)['cert'])
+        assert_true(cert_2_epoch_0 in self.nodes[0].getblock(self.nodes[0].getblockhash(height - 1), True)['cert'])
+        assert_true(cert_3_epoch_0 in self.nodes[0].getblock(mined, True)['cert'])
+
+        mark_logs("Checking certificates in mempool after block revertion(via invalidateblock)", self.nodes, DEBUG_MODE)
+        block_inv = self.nodes[0].getbestblockhash()
+        self.nodes[0].invalidateblock(block_inv)
+        sync_mempools(self.nodes[1:NUMB_OF_NODES])
+        sync_blocks(self.nodes[1:NUMB_OF_NODES])
+
+        # Checking certificate in mempool
+        assert_false(cert_1_epoch_0 in self.nodes[0].getrawmempool())
+        assert_false(cert_2_epoch_0 in self.nodes[0].getrawmempool())
+        assert_true(cert_3_epoch_0 in self.nodes[0].getrawmempool())
+
+        # Checking certificates in chain
+        assert_true(cert_1_epoch_0 in self.nodes[0].getblock(self.nodes[0].getbestblockhash(), True)['cert'])
+        assert_true(cert_2_epoch_0 in self.nodes[0].getblock(self.nodes[0].getbestblockhash(), True)['cert'])
+        assert_false(cert_3_epoch_0 in self.nodes[0].getblock(self.nodes[0].getbestblockhash(), True)['cert'])
+
+        # Check balance
+        bwt_address = self.nodes[1].getrawcertificate(cert_3_epoch_0, 1)['vout'][1]['scriptPubKey']['addresses'][0]
+        res = self.nodes[1].listtxesbyaddress(bwt_address)
+        assert_equal(res[0]['scid'], scid)
+        mark_logs("Height : {}".format(res[0]['vout'][1]['maturityHeight']), self.nodes, DEBUG_MODE)
+        ud = self.nodes[1].getunconfirmedtxdata(bwt_address)
+        assert_equal(bwt_amount_2, ud['bwtImmatureOutput'])
```

Comment:
This assert will fail because it is node 0 that invalidated the block; from node 1 perspective, nothing is changed and the topo quality cert is still cert3.

Reply:
Block invalidation and all checking moved to node 1.